### PR TITLE
[ISSUE #273 ]Add .NET Client to RocketMQ #273

### PR DESCRIPTION
**Pull request by ClementIV**
## What is the purpose of the change

Add .NET Client to RocketMQ 

## Brief changelog

1. Refactoring code structure
2. Publish the NuGet package
3. Write some documents
    * readme.md
   * Quick Start
   * API Reference

## Verifying this change

-[ISSUE #273 ] Add .NET Client to RocketMQ

### Add .NET Client
1. Refactoring code structure, see `readme.md` for more information.
2. Build a local `rocketmq` test environment
3. Connect `Aliyun RocketMQ` tested related APIs
4. Create a new project to test the NuGet package[rocketmq-client-dotnet](https://www.nuget.org/packages/rocketmq-client-dotnet/0.1.1/Manage)
5. Write some documents
    * readme.md
   * Quick Start
   * API Reference
### Unit Test

1. Add unit test
2. 43.22% coverage

### CODE REVIEW

File: rocketmq-client-dotnet/src/RocketMQ.NETClient/Interop/ConstValues.cs [hunk 1]
```diff
@@ -15,16 +15,19 @@
...
 
-namespace RocketMQ.Driver.Interop
+namespace RocketMQ.Client.Interop
 {
     public static class ConstValues
     {
-        public const string RocketMQDriverDllName = "rocketmq-client-cpp.dll";
+        public const string RocketMQDriverDllName = "CClient\\rocketmq-client-cpp.dll";
 
         public const string DiagnosticListenerName = "rocketmq-driver";
```
Comment (netcore-jroger) [inline rocketmq-client-dotnet/src/RocketMQ.NETClient/Interop/ConstValues.cs]: 以下四个常量这里面没有使用，如果不打算用，最好删除。

**Comment by netcore-jroger** [comment]
现在的：rocketmq-client-dotnet/src/`RocketMQ.NETClient`/
建议改成: rocketmq-client-dotnet/src/`RocketMQ.Client`/

File: rocketmq-client-dotnet/src/RocketMQ.NETClient/Consumer/MQPullConsumer.cs [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+using System.Threading.Tasks;
+
+namespace RocketMQ.Client.Consumer
+{
+    public class MQPullConsumer : IPullConsumer
+    {
+        #region default Options
+
+        private HandleRef _handleRef;
+        private readonly string LogPath = Environment.CurrentDirectory.ToString() + "\\PullConsumer_log.txt";
```
[the remark this answers is not in the document]
Comment (ClementIV) [reply rocketmq-client-dotnet/src/RocketMQ.NETClient/Consumer/MQPullConsumer.cs]: done

File: rocketmq-client-dotnet/src/RocketMQ.NETClient/Consumer/MQPushConsumer.cs [hunk 3]
```diff
@@ -15,35 +15,72 @@
...
-            
+
             this._handleRef = new HandleRef(this, handle);
+            this.SetPushConsumerLogPath(this.LogPath);
         }
-        
-        public IPushConsumerBuilder SetPushConsumerGroupId(string groupId)
+        public MQPushConsumer(string groupId)
+        {
+
```
[the remark this answers is not in the document]
Comment (ClementIV) [reply rocketmq-client-dotnet/src/RocketMQ.NETClient/Consumer/MQPushConsumer.cs]: done

File: rocketmq-client-dotnet/src/RocketMQ.NETClient/Consumer/MQPushConsumer.cs [hunk 4]
```diff
@@ -15,35 +15,72 @@
...
+            this.MQPushConsumerInit(groupId);
+        }
+
+        public MQPushConsumer(string groupId, string nameServerAddress)
+        {
+
```
[the remark this answers is not in the document]
Comment (ClementIV) [reply rocketmq-client-dotnet/src/RocketMQ.NETClient/Consumer/MQPushConsumer.cs]: done

File: rocketmq-client-dotnet/src/RocketMQ.NETClient/Consumer/PullConsumerWrap.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        public static extern int ReleaseSubscriptionMessageQueue(ref CMessageQueue mqs);
+
+        [DllImport(ConstValues.RocketMQDriverDllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern CPullResult Pull(HandleRef consumer, ref CMessageQueue mq, string subExpression,Int64 offset, int maxNums);
+
+        [DllImport(ConstValues.RocketMQDriverDllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern  int ReleasePullResult(CPullResult pullResult);
+      
+    }
+
```
[the remark this answers is not in the document]
Comment (ClementIV) [reply rocketmq-client-dotnet/src/RocketMQ.NETClient/Consumer/PullConsumerWrap.cs]: done

File: rocketmq-client-dotnet/src/RocketMQ.NETClient/Interop/ConstValues.cs (hunk 1 above)
Comment (ClementIV) [reply rocketmq-client-dotnet/src/RocketMQ.NETClient/Interop/ConstValues.cs]: Deleted

**Comment by John0King** [comment]
what happend ?

**Comment by vongosling** [comment]
@duhenglucky

**Pull request closed without merging**